### ISSUE TITLE
conflict in generated code if imported in multiple projects

### ISSUE THREAD

**Issue by CzBuCHi** [martinothamar/Mediator#62]
im unable to use `Mediator.SourceGenerator` in project, that is referencing different project which is already using `Mediator.SourceGenerator` - second project will throw at me this message: ([CS0436](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs0436))

`Error	CS0436	The type 'MediatorDependencyInjectionExtensions' in 'Mediator.SourceGenerator.Roslyn40\Mediator.SourceGenerator.IncrementalMediatorGenerator\Mediator.g.cs' conflicts with the imported type 'MediatorDependencyInjectionExtensions' in 'ProjectA'. Using the type defined in 'Mediator.SourceGenerator.Roslyn40\Mediator.SourceGenerator.IncrementalMediatorGenerator\Mediator.g.cs'.`

(technically it is 'only' warning, but i have TreatWarningsAsErrors enabled and therefore my build fails... )

affected types:
```
Mediator.MediatorOptions
Mediator.MediatorOptionsAttribute
Microsoft.Extensions.DependencyInjection.MediatorDependencyInjectionExtensions
```

ideas how to fix this:

- move `Mediator.MediatorOptions` and `Mediator.MediatorOptionsAttribute` to `Mediator.Abstractions` package 
      - they looks like they do not need to be generated
- generate `MediatorDependencyInjectionExtensions` in namespace defined by `MediatorOptions:Namespace` property
      - this way i can use multiple generated AddMediator() methods

-or-

make all 3 affected classes internal 
    - this will force to call AddMediator method from within project, that generated it but is simple fix

**Comment by martinothamar** [comment]
Thanks for reporting! To understand the usecase more, why is this dependency there? Are there two DI containers? one in each project? or can ProjectA run both independently and as part of the failing project (so that there is only one DI container)?

My initial feeling is that your second suggestion is the most appropriate in this case, but I'd like to understand the whole context. I never considered that there could be 2 projects having separate Mediator implementations in the same output binary basically.

**Comment by CzBuCHi** [comment]
hello, sorry for late reply

im experimenting with onion architecture / domain driven developement so my project structure is probably wrong ...

im using single DI provided by asp.net core.
my host project is basically `program.cs` from aspnet core minimal api template; modules are registered here
plan is to have separate modules, that can be enabled in host in random order
each module is using mediator internaly for CQRS and exporting extension methods to register itself in host project
module may be dependent on another module domain / use internal parts of that module

PS: now im also with te idea of making mediator classses internal - theyere implementation details after all ...

**Comment by michaeljfarr** [comment]
Hi. I think I have a related question, apologies if this isn't connected to this issue.

I am wanting to know how to use Mediator where multiple projects (separate assemblies) have Mediator handlers. Currently, it looks like there can be multiple "Mediator"s, there can only be one IMediator. Originally this wasn't a problem because I had all my handlers in the same project, but this has become cumbersome as the number of handlers and their interdependencies has grown. 

Please let me know if there is a straightforward way of achieving what I am describing. There is a good possibility that I've made a mistake with my project setup. The rest of this message only relates to the scenario where what I am trying to do isn't currently possible.

To explain the situation better, here is a simplified project layout I am working with:

Assembly1: ServiceGroupA (containing handlers)
Assembly2: ServiceGroupB (containing handlers)
Assembly3: Models
Assembly4: Program (an executable of some type)

ServiceGroupA and ServiceGroupB depend on "Models"; but each has 99% of its classes marked as internal.  No project dependency is allowed between ServiceGroupA and ServiceGroupB.
Program creates a service container including services both ServiceGroupA and ServiceGroupB (using a public extension method in each "AddServiceGroup[A|B]")

Program wants to do Mediator.Send(some IRequest type from "Models") and sometimes (rarely) one of the ServiceGroups will want to send a message to one another via the Mediator.  In either case, this should work out which ServiceGroup to send the request to. 

However, I see now that only one IMediator is registered, so there is only one Send and any messages sent to the second Service Group's handlers will respond with MissingMessageHandlerException.

I'd have expected that all mediators are registered in their own namespaces (implementing ISender).  Then the global IMediator implementation would send by looping over each Sender. Only if all of the sub-mediators failed to send would a MissingMessageHandlerException be thrown. I'd probably agree with CzBuCHi that MediatorOptions doesn't need to be generated.

Note: I am using 2.0.29.

**Comment by martinothamar** [comment]
So the SourceGenerator package is what is generating the options class, the DI extension method and the mediator implementation itself. It was meant to be used in the outermost package, typically where you use the generic host or ASP.NET Core to construct the runtime, DI container etc. 

To create queries, commands, requests and their handlers you only have to reference the abstractions package. So in the dependency graph from single output binary project (console, web, backgroundservice etc) there should only be 1 PackageReference and it should be at the outermost level (project file for the console, web or backgroundservice), everywhere else you should only need the abstractions package. 

This is how I've designet it so far, and I am using this in clean/onion architecture setups at work as well. On to your examples

From @CzBuCHi 
> each module is using mediator internaly for CQRS and exporting extension methods to register itself in host project
module may be dependent on another module domain / use internal parts of that module

As far as I can understand this should be fine, the module projects just depend on the abstractions package and inject `IMediator` or similar where necessary. The host project references the sourcegenerator and from there it will handle messages and their handles across modules. 

From @michaeljfarr 
>To explain the situation better, here is a simplified project layout I am working with:
>
>Assembly1: ServiceGroupA (containing handlers)
>Assembly2: ServiceGroupB (containing handlers)
>Assembly3: Models
>Assembly4: Program (an executable of some type)
>
>ServiceGroupA and ServiceGroupB depend on "Models"; but each has 99% of its classes marked as internal. No project >dependency is allowed between ServiceGroupA and ServiceGroupB.
>Program creates a service container including services both ServiceGroupA and ServiceGroupB (using a public extension >method in each "AddServiceGroup[A|B]")

Same as above, this seems to me like it should work. If only Assembly4|Program is referencing the SourceGenerator package, it should map out all the requests and handlers from there and it will know about all the other servicegroups. Note that a request from Models can't be handled from multiple service groups, then it would have to be a notification. But apart from this it doesn't seem like it should fail. 

To summarize, you should typically only have 1 outermost project referencing  the sourcegenerator package, and in a single DI container there should only be 1 AddMediator call, this library doesnt expect there to be multiple implementations of `IMediator` etc in the dependency graph. 

Hopefully I've understood your examples and clarified how this package works. If I'm wrong here please correct me. I will make sure to clarify this in the README as well as I'm not sure if this is specified anywhere apart from maybe the clean architecture sample in this repo.

Also related question in #71

**Comment by michaeljfarr** [comment]
Ah, ok - yes that does solve the issue I was having - thanks. 

It has meant that all my handlers need to be public accessible to the Program though. They are marked as internal, so I have just included an InternalsVisibleTo(Program) in Assembly1 and Assembly2. This is so the generated Mediator can add the services and associated RequestClassHandlerWrapper.

It would be slightly better if I could keep those handlers private within their respective Assemblies. However, it does seem more convenient to have one source generator. It is not a bother for the moment, my project still isn't big enough to worry too much about these kinds of things.

Thanks again.

**Comment by AntonC9018** [comment]
I don't think source generating in a single project is an acceptable solution. There may be a case where:
- One can't make handlers visible to the consumer;
- The consumer wants to selectively register handlers from specific projects.
Reading through the discussion, none of this seems to be possible.

**Comment by martinothamar** [comment]
Thanks for the feedback @AntonC9018 !

> * One can't make handlers visible to the consumer;

So Mediator has the same limitation here as any other C# code. Say if you were to implement a plain service in project B, mark it internal, and then try to call it from project A - you would get a compiler error (ignoring `InternalsVisibleToAttribute`), becaues you've just said that it should be internal. If you install Mediator in project A here, it has the exact same limitations as you would have trying to call the handlers manually yourself.

> * The consumer wants to selectively register handlers from specific projects.

I'm not sure what this means. Since for any request/command/query there should be _exactly 1_ handler, so we can't have multiple handlers for a single message (regardless of the location of the handlers).

What kind of architecture are you going for? Does it look like the one described by @michaeljfarr?


One idea that I've been thinking about that would let you stick with `internal` handlers would be to "shard" the Mediator implementations kind of by assembly. If both Mediator packages included source generators, where the current `Mediator.SourceGenerator` package kind of routed based on the message handler assembly.

```mermaid
flowchart TB
    IMediator-->IMediatorProxyA
    IMediator-->IMediatorProxyB
    subgraph Program
    IMediator
    end
    subgraph ServiceGroupB
    IMediatorProxyB-->HandlerB
    end
    subgraph Models
    Request0
    Request1
    end
    subgraph ServiceGroupA
    IMediatorProxyA-->HandlerA
    end
```

Request0..N can still just have 1 handler, so the "main" source generator would still have diagnostics based on that. But the DI container would cover all these modules and only the `IMediatorProxy`'s (and messages/requests) would have to be public

**Comment by AntonC9018** [comment]
@martinothamar Actually, I think you're right about the service group thing. To make the separate projects more modular, we have put the mediator registrations in a separate method in each project, then we call those methods for each project, from the main project. Each of those initialization methods registers all handlers from the executing assembly (it seems like we just call `services.AddMediatR(Assembly.GetExecutingAssembly())`). I just assumed each of these methods would add on to the central mediator registration, but it seems like I was wrong.

The benefit of modularly registering the handlers by assembly is that we can choose to not register handlers from certain projects from the main project.

Right now, I'm starting at the task of factoring out common logic into pipeline behaviors, so I will need to include the "common" assembly with the pipeline behaviors shared between all implementations for them to be resolved at registration time. Just including them once in their own registration method will not be enough.

So it will have to look kinda like this, graphically.

```mermaid
flowchart TB
    Main-->RegisterMediatorB
    Main-->RegisterMediatorA
    subgraph Program
    Main
    end
    subgraph ProjectA
    RegisterMediatorA-->HandlerA
    RegisterMediatorA-->HandlerB
    end
    subgraph ProjectB
    RegisterMediatorB-->HandlerC
    RegisterMediatorB-->HandlerD
    end
    RegisterMediatorA -->RegisterBehaviors
    RegisterMediatorB -->RegisterBehaviors
    subgraph Models
    RequestA
    RequestB
    RequestC
    RequestD
    end
    subgraph CommonBehaviors
    RegisterBehaviors-->PipelineBehaviorA
    end
```

I can't simplify it to just the following, because each registration is separate. 

```mermaid
flowchart TB
    Main-->RegisterMediatorB
    Main-->RegisterMediatorA
    subgraph Program
    Main
    end
    subgraph ProjectA
    RegisterMediatorA-->HandlerA
    RegisterMediatorA-->HandlerB
    end
    subgraph ProjectB
    RegisterMediatorB-->HandlerC
    RegisterMediatorB-->HandlerD
    end
    Main -->RegisterBehaviors
    subgraph Models
    RequestA
    RequestB
    RequestC
    RequestD
    end
    subgraph CommonBehaviors
    RegisterBehaviors-->PipelineBehaviorA
    end
```

**Comment by AntonC9018** [comment]
@martinothamar I came across a similar problem while writing my own source generator today. I have it construct a tree graph from all types in all projects, but the problem is that source generators run for each project separately. So now I'm thinking of ways I can process subgraphs instead and then join the results in a single root project. Also, can you explain at a high level how you managed to input the sources of all projects in a single source generator run, because, allegedly, it's impossible? I though it might give me some ideas.

**Comment by martinothamar** [comment]
Thanks for elaborating @AntonC9018 

> The benefit of modularly registering the handlers by assembly is that we can choose to not register handlers from certain projects from the main project.

Yeah so I think this is what's missing currently, that @CzBuCHi is also pointing out. I'll think about this in my 3.0 prototype, will need to come up with a way to make these methods easily consumable, since we now need to differentiate as you say.

> Also, can you explain at a high level how you managed to input the sources of all projects in a single source generator run, because, allegedly, it's impossible? I though it might give me some ideas.

Yeah the source generator is supposed to only care about the code in the "current" project, so the way that I've done this here is kind of hacky and not super-scalable (although I haven't run into any perf issues in my prod apps).

This is the relevant code: https://github.com/martinothamar/Mediator/blob/b1a560d06cf5021276891fc4612798e46b2d41af/src/Mediator.SourceGenerator.Implementation/Analysis/CompilationAnalyzer.cs#L288-L352

We fetch symbols from project/compilation references (that have a `Mediator` reference) recursively (within some limit). This is to map requests to handlers where they might be in separate projects. Again, this is pretty hacky and is not really intended usage from the source generator API, so I will be trying to get away from this model in 3.0.0. Current plan is to have source generation as part of every project, and let each project generate metadata in well known symbols (like a `MediatorSourceGenMeta` static class or something). I'm planning to start experimenting on this in the next couple of days hopefully

**Comment by mt89vein** [comment]
This problem also occurs, when you reference API/Worker (host project) in your Unit/Integration test project. Source generator package included only in host project, but leaked into test and generates duplicate types.

I found workaround - add to test csproj to disable source gen in it:

```
  <Target Name="DisableAnalyzers"
          BeforeTargets="CoreCompile">
    <ItemGroup>
      <Analyzer Remove="@(Analyzer)"
                Condition="'%(Filename)' == 'Mediator.SourceGenerator.Roslyn40'"/>
    </ItemGroup>
  </Target>
```
